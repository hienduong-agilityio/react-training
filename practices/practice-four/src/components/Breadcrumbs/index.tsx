// Libraries
import { Link } from 'react-router-dom';
import { memo } from 'react';

// SVG
import homeIcon from '@public/images/homeIcon.svg';
import arrowRight from '@public/images/arrowRightIcon.svg';

// Constants
import { ROUTE } from '@/constants';

export type BreadcrumbItem = {
  // id: The breadcrumb Items id
  id?: number;
  // label: The label to display for the breadcrumb item
  label?: string;
  // href: Optional URL for the breadcrumb item
  href?: string;
};

interface IBreadcrumbsProps {
  // items: Array of breadcrumb items
  items?: BreadcrumbItem[];
}

/**
 * Breadcrumbs component
 *
 * @returns {JSX.Element} Breadcrumbs element
 */
const Breadcrumbs = ({ items = [{ label: 'Home', href: ROUTE.ROOT }] }: IBreadcrumbsProps): JSX.Element => {
  return (
    <nav aria-label='breadcrumb'>
      <ul className='flex items-center space-x-1 md:space-x-2 rtl:space-x-reverse'>
        <li className='inline-flex items-center'>
          <Link to={ROUTE.ROOT} className='flex items-center text-sm font-medium text-gray-400 hover:text-primary-500'>
            <img src={homeIcon} alt='home' className='w-4 h-4 me-2.5 p-0' />
          </Link>
        </li>
        {items.map((item) => (
          <li key={item.label || item.id} className='inline-flex items-center'>
            <img src={arrowRight} alt='Arrow Right' className='rtl:rotate-180 w-6 h-6 text-gray-400 mx-1' />
            {item.href ? (
              <Link to={item.href} className='ms-1 text-sm font-medium text-gray-400 hover:text-primary-500 md:ms-2'>
                {item.label}
              </Link>
            ) : (
              <span className='ms-1 text-sm font-medium text-gray-400 md:ms-2'>{item.label}</span>
            )}
          </li>
        ))}
      </ul>
    </nav>
  );
};

export default memo(Breadcrumbs);
