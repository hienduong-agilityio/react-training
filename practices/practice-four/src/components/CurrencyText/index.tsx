export interface ICurrencyTextProps extends React.HTMLAttributes<HTMLParagraphElement> {
  // Currency: Optional estimated cost of the project, displayed as a financial estimate.
  currency?: number;
}

/**
 * CurrencyText component
 *
 * @returns {JSX.Element} - CurrencyItem element
 */
export const CurrencyText = ({ currency = 0, className = '' }: ICurrencyTextProps): JSX.Element => {
  return <p className={`${className}`}>{currency ? `US$ ${currency}k` : '-'}</p>;
};
