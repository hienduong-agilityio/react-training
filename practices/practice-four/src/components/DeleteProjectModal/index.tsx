// Components
import Modal from '@/components/common/Modal';
import Button from '@/components/common/Button';
import Spinner from '@/components/common/Spinner';

// Enums
import { BUTTON_COLORS, BUTTON_VARIANTS } from '@/enums/theme';

// Hooks
import { useProject } from '@/hooks/useProject';

// Store
import { ToastStore } from '@/stores/toastStore';

// Types
import type { IProjectItemProps } from '@/interfaces';

export interface IDeleteProjectModalProps {
  // projectId: The project id.
  projectId: string;
  // isModalOpen: A boolean indicating whether the modal is open.
  isModalOpen: boolean;
  // onCloseModal: A function to be called when the modal is requested to be closed.
  onCloseModal: () => void;
}

/**
 * The custom DeleteProjectModal component
 *
 * @returns {JSX.Element} - The DeleteProjectModal element
 */
export const DeleteProjectModal = ({
  projectId,
  isModalOpen = false,
  onCloseModal = () => {}
}: IDeleteProjectModalProps): JSX.Element => {
  const project = useProject({ id: projectId || '' });
  const { showToast } = ToastStore();

  const {
    mutate: { deleteProject },
    data: projectDetail,
    isMutating
  } = project;

  const projectName = (projectDetail as IProjectItemProps)?.projectName || 'this project';

  const handleDeleteProjectSuccess = () => {
    showToast('Project Deleted successfully', 'success');
    onCloseModal();
  };

  const handleDeleteProjectError = () => {
    showToast('Failed to deleted project', 'error');
    onCloseModal();
  };

  const handleDelete = () => {
    if (projectId) {
      deleteProject(projectId, {
        onSuccess: handleDeleteProjectSuccess,
        onError: handleDeleteProjectError
      });
    }
  };

  return (
    <Modal isOpen={isModalOpen} onClose={onCloseModal}>
      <div className='bg-light rounded-t-xl p-4 w-[440px]'>
        <p className='text-lg font-semibold text-gray-900 mb-3'>Delete Project</p>
        <p className='leading-6 text-gray-600'>
          Are you sure you want to delete <strong>{projectName}</strong>? If you delete, it will be permanently lost.
        </p>
      </div>
      <div className='bg-light rounded-b-xl flex justify-end gap-4 border-t-2 py-3'>
        <Button onClick={onCloseModal} variant={BUTTON_VARIANTS.OUTLINED}>
          Cancel
        </Button>
        <Button
          onClick={handleDelete}
          disabled={isMutating}
          variant={BUTTON_VARIANTS.CONTAINED}
          color={BUTTON_COLORS.DANGER}
          customClasses='mr-5'
        >
          {isMutating ? <Spinner /> : 'Delete'}
        </Button>
      </div>
    </Modal>
  );
};
