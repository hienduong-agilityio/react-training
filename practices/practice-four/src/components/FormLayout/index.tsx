interface FormLayoutProps {
  title: string;
  subtitle?: string;
  children: React.ReactNode;
}

export const FormLayout = ({ title, subtitle, children }: FormLayoutProps) => (
  <section>
    <div className='flex rounded-t-xl items-center bg-white justify-between p-4'>
      <p className='text-xl font-semibold text-gray-900'>{title}</p>
    </div>
    {subtitle && <p className='text-gray-600 mt-1 mb-4'>{subtitle}</p>}
    {children}
  </section>
);
