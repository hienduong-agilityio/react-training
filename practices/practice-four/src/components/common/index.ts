export { default as Avatar } from './Avatar';
export { default as Button } from './Button';
export { default as InputField } from './InputField';
export { default as Modal } from './Modal';
export { default as Pagination } from './Pagination';
export { default as SelectField } from './SelectField';
export * from './SkeletonLoader';
export { default as Spinner } from './Spinner';
export { default as StatusLabel } from './StatusLabel';
export { default as Tag } from './Tag';
export { default as Toast } from './ToastMessage';
