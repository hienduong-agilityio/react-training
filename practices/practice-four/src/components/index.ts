export { default as Breadcrumbs } from './Breadcrumbs';
export * from './CurrencyText';
export * from './DeleteProjectModal';
export * from './ErrorBoundary';
export * from './ErrorDisplay';
export * from './ProjectDetailSkeleton';
export { default as ProjectFilterDropdown } from './ProjectFilterDropdown';
export * from './ProjectFilterDropdown';
export { default as ProjectItem } from './ProjectItem';
export * from './ProjectItem';
export * from './ProjectOptionsDropdown';
export * from './ProjectTable';
export { default as ResourceTag } from './ResourcesTag';
export { default as SearchBox } from './SearchBox';
export * from './SuspenseLoader';
export { default as Timeline } from './Timeline';
export * from './ToastContainer';
export * from './FormLayout';
export * from './ProjectFormFields';
export * from './SubmitForm';

// Common
export * from './common';
