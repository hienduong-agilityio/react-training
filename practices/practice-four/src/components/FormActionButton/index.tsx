// Libraries
import { useFormStatus } from 'react-dom';

// Components
import { Button, Spinner } from '@/components';

// Enums
import { BUTTON_COLORS, BUTTON_SIZES, BUTTON_VARIANTS } from '@/enums';

interface IFormActionButtonProps {
  children: React.ReactNode;
  type?: 'submit' | 'button';
  color?: BUTTON_COLORS;
  variant?: BUTTON_VARIANTS;
  size?: BUTTON_SIZES;
  disabledOnPending?: boolean;
  showSpinner?: boolean;
  onClick?: () => void;
  customClass?: string;
}

export const FormActionButton = ({
  children,
  type = 'submit',
  variant = BUTTON_VARIANTS.CONTAINED,
  size = BUTTON_SIZES.LARGE,
  color = BUTTON_COLORS.PRIMARY,
  disabledOnPending = true,
  showSpinner = true,
  customClass = '',
  onClick
}: IFormActionButtonProps) => {
  const { pending } = useFormStatus();

  return (
    <Button
      type={type}
      disabled={disabledOnPending && pending}
      variant={variant}
      size={size}
      color={color}
      onClick={onClick}
      customClasses={customClass}
    >
      {pending && showSpinner ? <Spinner /> : children}
    </Button>
  );
};
