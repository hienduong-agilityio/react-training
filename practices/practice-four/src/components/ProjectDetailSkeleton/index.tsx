import { SkeletonLoader } from '@/components';

/**
 * ProjectDetailSkeleton components
 */
export const ProjectDetailSkeleton = () => {
  return (
    <div className='flex items-center p-8 h-full gap-36 space-y-6 flex-row'>
      <div className='flex-shrink-0 w-5/12 h-full flex flex-col items-center'>
        {/* Placeholder for Avatar */}
        <SkeletonLoader width='75%' height='300px' customClassName='mb-4' />
        <SkeletonLoader width='50%' height='20px' />
      </div>

      <div className='flex-grow lg:w-7/12 h-full space-y-6'>
        <SkeletonLoader width='70%' height='30px' customClassName='mb-4' />
        <SkeletonLoader width='20%' height='20px' customClassName='mb-2' />
        <SkeletonLoader width='100px' height='30px' customClassName='mb-4 rounded-md' />

        <SkeletonLoader width='40%' height='20px' customClassName='mb-2' />
        <SkeletonLoader width='60%' height='20px' customClassName='mb-4' />

        <SkeletonLoader width='40%' height='20px' customClassName='mb-2' />
        <SkeletonLoader width='60%' height='20px' customClassName='mb-4' />

        <SkeletonLoader width='40%' height='20px' customClassName='mb-2' />
        <SkeletonLoader width='100px' height='20px' customClassName='mb-4' />

        <SkeletonLoader width='40%' height='20px' customClassName='mb-2' />
        <SkeletonLoader width='100px' height='20px' />
      </div>
    </div>
  );
};
