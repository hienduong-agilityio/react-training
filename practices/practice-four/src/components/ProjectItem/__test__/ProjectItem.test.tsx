// Libraries
import { fireEvent, render, screen } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

// Components
import { ProjectItem, IProjectItem } from '@/components';

// Enums
import { STATUS } from '@/enums/status';

// Mock data for ProjectItem
const mockProjectItem: IProjectItem = {
  id: '123456',
  projectName: 'New Project',
  status: 'On hold',
  managerId: 1,
  lastUpdate: '2024-10-10',
  resources: '5',
  timeline: {
    timeStart: '2024-01-01',
    timeEnd: '2024-12-31'
  },
  budget: 10000,
  manager: {
    id: '1',
    managerName: 'John Doe',
    managerImage: '/path/to/image.jpg'
  },
  onDeleteProject: jest.fn()
};

describe('ProjectItem Component', () => {
  const renderComponent = (props: Partial<IProjectItem> = {}) => {
    return render(
      <BrowserRouter>
        <ProjectItem {...mockProjectItem} {...props} />
      </BrowserRouter>
    );
  };

  it('should render with provided props and match snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it('should display the correct project details', () => {
    renderComponent();

    expect(screen.getByText('456')).toBeInTheDocument();
    expect(screen.getByText('New Project')).toBeInTheDocument();
    expect(screen.getByText('N/A')).toBeInTheDocument();
    expect(screen.getByText('On hold')).toBeInTheDocument();
    expect(screen.getByText('2024-10-10')).toBeInTheDocument();
    expect(screen.getByText('5')).toBeInTheDocument();
    expect(screen.getByText(/\$?10000/)).toBeInTheDocument();
  });

  it('should use default props is provided', () => {
    renderComponent({});

    expect(screen.getByText('New Project')).toBeInTheDocument();
    expect(screen.getByText(STATUS.ON_HOLD)).toBeInTheDocument();

    const idCell = screen.getAllByRole('cell')[0];
    expect(idCell).toBeInTheDocument();

    const timelineStart = screen.queryByText('2024-01-01');
    const timelineEnd = screen.queryByText('2024-12-31');

    expect(timelineStart).not.toBeInTheDocument();
    expect(timelineEnd).not.toBeInTheDocument();
  });

  it('should trigger onDeleteProject when the delete option is clicked', () => {
    const mockOnDeleteProject = jest.fn();
    renderComponent({ onDeleteProject: mockOnDeleteProject });

    const row = screen.getByText('New Project').closest('tr');
    fireEvent.mouseEnter(row!);

    const dropdownButton = screen.getByRole('button');
    fireEvent.click(dropdownButton);
    const deleteOption = screen.getByText('Delete');
    fireEvent.click(deleteOption);

    expect(mockOnDeleteProject).toHaveBeenCalledWith('123456');
  });
});
