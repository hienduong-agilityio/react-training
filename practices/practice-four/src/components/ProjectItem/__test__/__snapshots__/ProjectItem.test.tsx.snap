// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectItem Component should render with provided props and match snapshot 1`] = `
<div>
  <tr
    class="border-b border-gray-200 hover:bg-light group"
  >
    <td
      class="px-3 py-5 font-medium"
    >
      <p>
        456
      </p>
    </td>
    <td>
      <a
        class="text-primary-500 font-semibold hover:underline cursor-pointer w-max"
        href="/projects/123456"
      >
        New Project
      </a>
    </td>
    <td
      class="text-center px-2"
    >
      <div
        class="inline-flex border-2 border-gray-100 items-center justify-center overflow-hidden bg-primary-0 rounded-lg w-7 h-7"
      >
        <span
          class="font-bold text-xs text-primary-500"
        >
          N/A
        </span>
      </div>
    </td>
    <td>
      <div
        class="w-max"
      >
        <span
          class="text-xs py-[1px] px-2 inline-flex gap-[6px] rounded font-medium leading-[18px] tracking-[0.36px] flex items-center justify-center color-default"
        >
          <svg
            class="inline-flex color-default"
            fill="none"
            height="6"
            role="img"
            viewBox="0 0 6 6"
            width="6"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              height="6"
              rx="2"
              width="6"
            />
          </svg>
          On hold
        </span>
      </div>
    </td>
    <td>
      <p
        class="text-gray-600 w-max"
      >
        2024-10-10
      </p>
    </td>
    <td
      class="text-center px-2"
    >
      <div>
        <span
          class="py-[3px] px-2 inline-flex gap-[6px] rounded font-medium leading-[18px] tracking-[0.36px] flex items-center justify-center color-default"
        >
          5
        </span>
      </div>
    </td>
    <td>
      <div
        class="flex gap-1 items-center"
      >
        <span
          class="w-[98px] h-5 inline-flex gap-[6px] rounded font-medium leading-[18px] tracking-[0.36px] flex items-center justify-center color-default"
        >
          -
        </span>
        <svg
          aria-label="Next"
          fill="none"
          height="16"
          role="img"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M6.5 11L9.5 8L6.5 5"
            stroke="#868FA0"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
        </svg>
        <span
          class="w-[98px] h-5 inline-flex gap-[6px] rounded font-medium leading-[18px] tracking-[0.36px] flex items-center justify-center color-default"
        >
          -
        </span>
      </div>
    </td>
    <td>
      <div
        class="flex items-center w-44 justify-between"
      >
        <p
          class="text-primary-400 font-bold"
        >
          US$ 10000k
        </p>
        <div
          class="group-hover:block hidden"
        >
          <div
            class="relative inline-block text-left group"
          >
            <button
              class="flex gap-2 items-center focus:ring-4 focus:ring-primary-100 rounded-md cursor-pointer disabled:opacity-45 border-white text-primary-400 hover:border-2 hover:border-primary-100 px-4 py-2 text-base text-black hover:border-none hover:bg-gray-100"
            >
              <img
                alt="Menu"
                src=""
              />
            </button>
          </div>
        </div>
      </div>
    </td>
  </tr>
</div>
`;
